src/main/java/gregtech/common/covers/CoverFluidRegulator.java

```diff
@@ -0,0 +1,245 @@
+            if (sourceFluid == null || sourceFluid.amount == 0 || !fluidFilter.test(sourceFluid)) continue;
+            sourceFluid.amount = supplyAmount;
+            sourceFluid = sourceHandler.drain(sourceFluid, false);
+            if (sourceFluid == null || sourceFluid.amount != supplyAmount) continue;
+            int canInsertAmount = destHandler.fill(sourceFluid, false);
+            if (canInsertAmount == supplyAmount) {
+                sourceFluid = sourceHandler.drain(sourceFluid, true);
+                if (sourceFluid != null && sourceFluid.amount > 0) {
+                    destHandler.fill(sourceFluid, true);
+
+                    fluidLeftToTransfer -= sourceFluid.amount;
+                    if (fluidLeftToTransfer == 0) break;
+                }
+            }
+        }
+        return transferLimit - fluidLeftToTransfer;
+    }
+
+    protected int doKeepExact(int transferLimit, IFluidHandler sourceHandler, IFluidHandler destHandler, Predicate<FluidStack> fluidFilter, int keepAmount) {
+        int fluidLeftToTransfer = transferLimit;
+        for (IFluidTankProperties tankProperties : sourceHandler.getTankProperties()) {
+            FluidStack sourceFluid = tankProperties.getContents();
+            if (sourceFluid == null || sourceFluid.amount == 0 || !fluidFilter.test(sourceFluid)) continue;
+            sourceFluid.amount = keepAmount;
+            FluidStack destFluid = destHandler.drain(sourceFluid, false);
+            int amountToDrainAndFill;
+            //no fluid in destination
+            if (destFluid == null) {
+                amountToDrainAndFill = Math.min(keepAmount, fluidLeftToTransfer);
+            //if the amount of fluid in the destination is sufficient or the destinations fluid isnt equal to the sources
+            //how to check if destHandler is full?
+            } else if (destFluid.amount >= keepAmount || !destFluid.isFluidEqual(sourceFluid)) {
+                continue;
+            } else {
+            //if keepAmount is larger than the transferLimit we will have to stock it over several ticks (seconds?)
+                amountToDrainAndFill = Math.min(keepAmount - destFluid.amount, fluidLeftToTransfer);
+            }
+            sourceFluid.amount = amountToDrainAndFill;
+            sourceFluid = sourceHandler.drain(sourceFluid, false);
+            if (sourceFluid == null || sourceFluid.amount != amountToDrainAndFill) continue;
+            int canInsertAmount = destHandler.fill(sourceFluid, false);
+            if (canInsertAmount == amountToDrainAndFill) {
+                sourceFluid = sourceHandler.drain(sourceFluid, true);
+                if (sourceFluid != null && sourceFluid.amount > 0) {
+                    destHandler.fill(sourceFluid, true);
+
+                    fluidLeftToTransfer -= sourceFluid.amount;
+                    if (fluidLeftToTransfer == 0) break;
+                }
+            }
+        }
+        return transferLimit - fluidLeftToTransfer;
+    }
```

Comment:
These are both quite similar would be possible to extract some parts to methods and pass in just values. Possible place them to GTFluidUtils?

Reply:
It might be possible to extract the common parts of these but after spending some time looking at the code no obvious, easy to read solution comes to me. I could move them to GTFluidUtils. Currently I followed the same workflow as in CoverRoboticArm which keeps all these methods in the same class.

Reply:
I have marked place which I feel is extractable via another comment. Also I feel we should leverage `GTFluidUtils` as much as is reasonable because we are manipulating fluids from quite few places